### DISCUSSION TITLE
How to store custom ID?

### DISCUSSION THREAD

**Discussion by SwapnilSoni1999** [prisma/prisma#20865] [Q&A] [answered]
### Question

Hi I'm pretty new to prisma,
I am using postgresql and I want to store ID as custom prefix such as `SK00XXX` something like this 
So whenever I add user it should store as `SK00001` and incrementing value

**Comment by ludralph** [comment, marked as answer]
Hi @SwapnilSoni1999 👋 

Thank you for raising this question.

Currently, Prisma does not have support for creating custom IDs in the format you have specified out of the box. We do have a feature request to support custom user-defined ID patterns.

- https://github.com/prisma/prisma/issues/6719

However, as a workaround based on this [comment](https://github.com/prisma/prisma/issues/6719#issuecomment-1383678506) in the issue, you can leverage [dbgenerated()](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#dbgenerated) to create a database function that generates the custom `ID` you want. 
For example:

1. Create a Prisma schema below:

```prisma
model Product {
  id    String    @id @default(dbgenerated("generate_custom_id('SK'::text, 1::smallint, 5::smallint)"))
  name String
}
```

2. Run `npx prisma migrate dev --create-only` to generate a new migration without applying it.

3. Paste this SQL at the top of the generated migration file
```sql
CREATE OR REPLACE FUNCTION generate_custom_id(prefix text, start_value smallint, padding_length smallint) RETURNS text AS $$
DECLARE
    new_value smallint;
    formatted_id text;
BEGIN
    SELECT max(CAST(SUBSTRING(id, length(prefix) + 1) AS smallint)) INTO new_value FROM "Product"; -- Change "Product" to the actual table name
    
    IF new_value IS NULL THEN
        new_value := start_value;
    ELSE
        new_value := new_value + 1;
    END IF;
    
    formatted_id := prefix || lpad(new_value::text, padding_length, '0');
    
    RETURN formatted_id;
END;
$$ LANGUAGE PLPGSQL VOLATILE;
```
4. Run `npx prisma migrate dev` to finish creating the migration.

5. You can now go ahead and create a new `Product` like:
```js
const res = await prisma.product.create({
  data:{
    name:"product 1",
  }
})
```
The output looks like:
```js
{ id: 'SK00001', name: 'product 1' }
```

After creating multiple products, you should now see that the custom ID is generated and incremented as expected.
For example:
```js
[
  { id: 'SK00001', name: 'product 1' },
  { id: 'SK00002', name: 'product 2' },
  { id: 'SK00003', name: 'product 3' }
]
```
 
If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by ludralph** [reply]
Hi @SwapnilSoni1999 

Do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Reply by SwapnilSoni1999** [reply]
Amazing! Thankyou so much! <3 its working fine.

**Reply by mr-mern** [reply]
@ludralph  can you add code snippet of product model .. it will make better understanding how this function is bind with product.id

**Reply by ludralph** [reply]
Hi @hassan-debug, This is the product model I used. It is stated above as well.
```prisma
model Product {
  id    String    @id @default(dbgenerated("generate_custom_id('SK'::text, 1::smallint, 5::smallint)"))
  name String
}
```

**Comment by devSajan** [comment]
what if i want to use MongoDB object id in string format in the Postgres id field and I also want to add a validation, if it's failing to match the object id regex we should get an error. can we achieve that?

**Comment by Rakeshpradhan1999** [comment]
```
`CREATE OR REPLACE FUNCTION generate_custom_id(prefix text, start_value smallint, padding_length smallint, table_name text) RETURNS text AS $$
DECLARE
    new_value smallint;
    formatted_id text;
BEGIN
    SELECT max(CAST(SUBSTRING(id, length(prefix) + 1) AS smallint)) INTO new_value FROM table_name; 
    
    IF new_value IS NULL THEN
        new_value := start_value;
    ELSE
        new_value := new_value + 1;
    END IF;
    
    formatted_id := prefix || lpad(new_value::text, padding_length, '0');
    
    RETURN formatted_id;
END;
$$ LANGUAGE PLPGSQL VOLATILE;`
```

**Reply by Rakeshpradhan1999** [reply]
I tried this making the table_name dynamic so we can use it dynamically but its not working while creating data
![image](https://github.com/user-attachments/assets/35352cd3-1456-4b42-9bc0-de59c78737f9)
